File: 22/22.clj
(def r #".*-x(\d+)-y(\d+)\s+\d+T\s+(\d+)T\s+(\d+)T.*")

(defn parse-line [line]
  (let [[_ x y used avail] (re-matches r line)]
    { :x (Long/parseLong x)
      :y (Long/parseLong y)
      :used (Long/parseLong used)
      :avail (Long/parseLong avail) }))

(defn parse [filename]
  (->> (slurp filename)
       clojure.string/split-lines
       rest ; df command
       rest ; df header
       (map parse-line)))

(defn viable-pair? [a b]
  (and (pos? (:used a))
       (and (not= (:x a) (:x b)) (not= (:y a) (:y b)))
       (<= (:used a) (:avail b))))

(defn viable-pairs [nodes node]
  (filter (partial viable-pair? node) nodes))

(defn count-viable-pairs [nodes]
  (count (map (partial viable-pairs nodes) nodes)))

(let [nodes (vec (parse "input.txt"))]
  (println (first nodes) (last nodes))
  (println (count-viable-pairs nodes)))

File: 22/22.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

// The initial empty one ('_') has 0 usage
// The blockers ('#') have > 90T usage
// Others are free ('.')

/*
First, get the empty space from its initial starting point,
around the immovable wall, and up next to the target data:

*.........8901234567890123456789G
..........7......................
..........6......................
..........5......................
..........4......................
..........3......................
..........2......................
..........1......................
..........0......................
..........9######################
..........8......................
..........7......................
..........6543210987654321_......
.................................
...etc

The 49th move has left us in this state for the top 2 rows:

*.............................._G
.................................

We move G one step to the left. This takes 5 moves:

init   1    2    3    4    5
._G  .G_  .Gc  .Gc  .Gc  _Gc
abc  abc  ab_  a_b  _ab  .ab

After repeating this process 31 times, the top-left looks like this:

_G..
....

Total moves: 49 + 31*5 = 204.

And the last move to put G in the upper left => 205 moves.
*/

namespace Advent { class Problem22_Visualization {

    static Regex r = new Regex(
        @"^/dev/grid/node-x(\d+)-y(\d+)\s+\d+T\s+(\d+)T\s+\d+T\s+\d+%$");

    static void ParseAndPrint(string[] lines) {
        Console.WriteLine(lines.Length);
        var grid = new char[33,30];
        foreach (var line in lines.Skip(2)) {
            var m = r.Match(line);
            var x = Int32.Parse(m.Groups[1].Value);
            var y = Int32.Parse(m.Groups[2].Value);
            var used = Int32.Parse(m.Groups[3].Value);

            var status = used == 0
                ? '@'
                : used > 90 ? '#' : '.';

            grid[x,y] = status;
        }

        for (int y = 0; y < 30; y++) {
            for (int x = 0; x < 33; x++) {
                Console.Write(grid[x,y]);
            }
            Console.WriteLine();
        }
    }

    static void Main() {
        ParseAndPrint(System.IO.File.ReadAllLines("input.txt"));
    }
}}